### PR TITLE
Add SpectatorHistogram extension

### PR DESCRIPTION
**Pull request by bsyk**
### Description

Adds the contribution extension providing support for SpectatorHistogram. A fast, small alternative to data-sketches or T-Digest for computing approximate percentiles.

See documentation included in the PR for more details.

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by adarshsanjeev** [comment]
Thanks for the PR! This looks like a really cool addition to Druid. I'm going through the PR, and will add comments after it's done.

**Review by suneet-s — COMMENTED** [review]
Thanks for the contribution @bsyk !

I have not reviewed the code in depth, and have mostly read through the documentation.

From skimming the code, I think the extension is added in a safe way, and can be improved on in future patches.

Some things that could be done in the future:
* Add vectorized implementations of the aggregator so that it allows queries to leverage that functionality.
* Add sql bindings

File: extensions-contrib/spectator-histogram/src/main/java/org/apache/druid/spectator/histogram/NullableOffsetsHeader.java
Comment (adarshsanjeev) [inline extensions-contrib/spectator-histogram/src/main/java/org/apache/druid/spectator/histogram/NullableOffsetsHeader.java:38]: nit: A javadoc for this class as a small summary of its usages would help make the code more readable

File: extensions-contrib/spectator-histogram/src/main/java/org/apache/druid/spectator/histogram/SpectatorHistogramAggregatorFactory.java
Comment (adarshsanjeev) [inline extensions-contrib/spectator-histogram/src/main/java/org/apache/druid/spectator/histogram/SpectatorHistogramAggregatorFactory.java:270]: Also to be resolved

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (bsyk) [reply docs/development/extensions-contrib/spectator-histogram.md:243]: Where only a single percentile is wanted, often median or 95th, it's slightly nicer to get a single value back, rather than having to extract from an array in the results.
Also not a strong opinion.

Is the note misleading? It's trying to say, "if you want multiple percentiles from the same underlying metric, then ask for them all at once, rather than as separate metrics". 1 query being more efficient than 2.

File: extensions-contrib/spectator-histogram/src/main/java/org/apache/druid/spectator/histogram/SpectatorHistogramAggregatorFactory.java
Comment (bsyk) [reply extensions-contrib/spectator-histogram/src/main/java/org/apache/druid/spectator/histogram/SpectatorHistogramAggregatorFactory.java:270]: This was left over from an earlier implementation and no longer relevant.

**Review by adarshsanjeev — COMMENTED** [review]
Thanks for addressing the comments! I've gone through the changes, and they seem to be fine. The build failures look unrelated.

**Review by adarshsanjeev — APPROVED** [review]

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:38]:
```suggestion
```
Combine this into the section above. Note that the `>` notation actually makes text smaller and italicized, which may get overlooked.

Example: https://druid.apache.org/docs/latest/operations/durable-storage/ ("Note that only S3 is supported as a durable storage location.")

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:48]: Capitalize "Wikipedia" when used in text (here and elsewhere)

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [reply docs/development/extensions-contrib/spectator-histogram.md:45]: Add new line before and after figures

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:48]:
```suggestion
As rollup improves, so does the size savings. For example, when you ingest the Wikipedia dataset
```

Comment (vtlim) [reply docs/development/extensions-contrib/spectator-histogram.md:48]: Prefer "you" instead of "we" in docs. https://github.com/apache/druid/blob/master/docs/development/docs-contribute.md#style-checklist

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:53]:
```suggestion
with day-grain query granularity and remove all dimensions except `countryName`,
this results in a segment that has just 106 rows. The base segment has 87 bytes per row.
Compare the following bytes per row for SpectatorHistogram versus DataSketches:
* An additional `spectatorHistogram` column adds 27 bytes per row on average.
* An additional `quantilesDoublesSketch` column adds 255 bytes per row.

SpectatorHistogram reduces the additional storage size by 9.4 times in this example.
Storage gains will differ per dataset depending on the variance and rollup of the data.
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:68]: Refer to "DataSketch" in line with the docs (here and elsewhere)
https://druid.apache.org/docs/latest/development/extensions-core/datasketches-extension/

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:69]: What does "opinionated" mean in this context?

Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:69]:
```suggestion
opinionated and optimized for typical measurements from cloud services and web apps.
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:70]:
```suggestion
For example, measurements such as page load time, transferred bytes, response time, and request latency.
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:71]:
```suggestion
Through some trade-offs SpectatorHistogram provides a significantly more compact
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:73]:
```suggestion
data-sketches. Note that results depend on the dataset.
Also see the [limitations](#limitations] of this extension.
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:81]:
```suggestion
* Supports positive long integer values within the range of [0, 2^53). Negatives are
coerced to 0.
* Does not support decimals.
* Does not support Druid SQL queries, only native queries.
* Does not support vectorized queries.
* Generates 276 fixed buckets with increasing bucket widths. In practice, the observed error of computed percentiles ranges from 0.1% to 3%, exclusive. See [Bucket boundaries](#histogram-bucket-boundaries) for the full list of bucket boundaries.
```
Start each bullet in the same way (verb or noun)

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:84]:
```suggestion
:::tip
If these limitations don't work for your use case, then use [DataSketches](../extensions-core/datasketches-extension.md) instead.
:::
```
Uses Docusaurus-style admonitions to capture attention https://docusaurus.io/docs/2.x/markdown-features/admonitions

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:87]:
```suggestion
The SpectatorHistogram aggregator can generate histograms from raw numeric
```
Prefer active voice

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:88]:
```suggestion
values as well as aggregating or combining pre-aggregated histograms generated using
```
Use "or" or "and" instead of "/" which can be ambiguous

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:93]:
```suggestion
incorrect to generate histograms from already rolled-up summed data.
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:96]:
```suggestion
`percentilesSpectatorHistogram` (plural), to compute approximate 
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:103]:
```suggestion
and accuracy are comparable. However, the DataSketch aggregator supports negative values,
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:104]: It seems DataSketches also need to be loaded. Did you mean "download"?
https://druid.apache.org/docs/latest/development/extensions-core/datasketches-extension/
```
druid.extensions.loadList=["druid-datasketches"]
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:120]:
```suggestion
histogram. The keys need not be ordered or contiguous. For example:
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:141]:
```suggestion
See [Histogram bucket boundaries](#histogram-bucket-boundaries) for the full list of bucket boundaries.
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:161]:
```suggestion
underlying implementation. If you use the Atlas-Druid service, the different types
signal the service on how to handle the resulting data from a query.
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:245]:
```suggestion
> It's more efficient to request multiple percentiles in a single query
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:252]:
```suggestion
The results contain arrays matching the length and order of the requested
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:271]: We don't typically include an appendix in docs. Consider making this "Examples" and promoting the heading for bucket boundaries.
```suggestion
## Examples
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:400]:
```suggestion
## Histogram bucket boundaries
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:401]:
```suggestion
The following array lists the upper bounds of each bucket index. There are 276 buckets in total.
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:403]:
```suggestion
The bucket widths increase as the bucket index increases. This leads to a greater absolute error for larger values, but maintains a relative error of rough percentage across the number range.
```

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:404]:
```suggestion
For example, the maximum error at value 10 is zero since the bucket width is 1 (the difference of 11-10). For a value of 16,000,000,000, the bucket width is 1,431,655,768 giving an error of up to ~8.9%. In practice, the observed error of computed percentiles is in the range of (0.1%, 3%).
```
Not sure if it's 11-10 or 10-9. Also it's not clear how you get 8.9%. Consider explaining in more detail, or linking to relevant docs.

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (bsyk) [reply docs/development/extensions-contrib/spectator-histogram.md]: There are opinions built into the implementation, the main one being that smaller values should be more accurately recorded. The opinion that we can afford to be off by a few million once we're into the many-millions range, but don't want to be off by more than 1 at the very small numbers.
Happy to try to rephrase.

File: docs/development/extensions-contrib/spectator-histogram.md
Comment (vtlim) [inline docs/development/extensions-contrib/spectator-histogram.md:68]:
```suggestion
It provides similar functionality to the built-in DataSketches `quantilesDoublesSketch` aggregator, but is
opinionated to maintain higher accuracy at smaller values.
See [Bucket boundaries](#histogram-bucket-boundaries) for more information and an example.
The SpectatorHistogram is optimized for typical measurements from cloud services and web apps,
such as measurements such as page load time, transferred bytes, response time, and request latency.

```

**Review by vtlim — APPROVED** [review]
Left one more comment to try to explain the opinionated behavior. Otherwise LGTM for docs.

**Comment by bsyk** [comment]
> Left one more comment to try to explain the opinionated behavior. Otherwise LGTM for docs.

Fantastic. Thanks for all your suggestions, they've made the docs a lot more clear.

**Review by maytasm — APPROVED** [review]

**Comment by maytasm** [comment]
@suneet-s @vtlim @adarshsanjeev 
Are we all good on merging this change in? I'll merge this change in at the end of the day if no one has any objection.
This change is contained in it's own extension (and is a contrib extension) so should be safe to merge.
Including my approval, we have 3 +1s from committers and total reviews from 4 committers. 
I think this extension does serve a specific use case and does it well. This could be beneficial to some people and hence I do see a value in getting this merge in so that Druid users can easily use/try it out. 
By getting this extension merge in, we can also keep this extension up-to-dated with any new Druid changes, encourage improvements to this extension from the community (such as adding vectorization, etc) and get feedback about the extension. Note that at Netflix, we have been running this extension at scale in production for a few years so it is also battle tested.

**Comment by suneet-s** [comment]
@maytasm Good with me to merge once CI is green. Thanks @bsyk for the contribution and patience with the review :)

File: extensions-contrib/spectator-histogram/src/main/java/org/apache/druid/spectator/histogram/SpectatorHistogramAggregatorFactory.java
Comment (clintropolis) [inline extensions-contrib/spectator-histogram/src/main/java/org/apache/druid/spectator/histogram/SpectatorHistogramAggregatorFactory.java:209]: these methods are deprecated and will be removed at some point, please implement `getIntermediateType` and `getResultType` instead.

**Review by clintropolis — COMMENTED** [review]
since this is already approved is probably fine to do the changes i suggested as follow-up PR

**Comment by maytasm** [comment]
@bsyk 
Thank you for the change!
Merging this in as CI failure is unrelated to this PR (and master branch is failing on the same failure)

**Pull request merged**
